### PR TITLE
vm: implement new PUSH opcodes

### LINKED ISSUE #873: Implement neo-vm 3 PUSH opcode set
Change PUSH opcodes to neo-project/neo-vm#235 (neo-project/neo-vm#247). Make all the code use them and remove some code duplication spotted in #869 (script generators should use `emit.Bytes()`).

### PR DESCRIPTION
**Pull request by fyrchik**
Closes #873 .

### CODE REVIEW

File: pkg/vm/emit/emit.go [hunk 1]
```diff
@@ -32,20 +33,35 @@ func Bool(w *io.BinWriter, ok bool) {
...
+	case i >= 0 && i < 16:
 		val := opcode.Opcode(int(opcode.PUSH1) - 1 + int(i))
 		Opcode(w, val)
 	default:
-		bInt := big.NewInt(i)
-		val := IntToBytes(bInt)
-		Bytes(w, val)
+		buf := IntToBytes(big.NewInt(i))
+		l := len(buf)
+		if l > 32 {
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:56]: I don't think it could happen with `int64`.

File: pkg/vm/emit/emit.go [hunk 2]
```diff
@@ -32,20 +33,35 @@ func Bool(w *io.BinWriter, ok bool) {
...
+			w.Err = errors.New("too big integer")
+			return
+		}
+
+		// l != 0 becase of switch
+		padSize := byte(8 - bits.LeadingZeros8(byte(l-1)))
+		Opcode(w, opcode.Opcode(padSize))
+		w.WriteBytes(padRight(1<<padSize, buf))
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:64]: Creating yet another buffer in `padRight` doesn't look right. We can either use some `[32]byte` here (as we know that it won't exceed this length) and copy to/slice from it or just enlarge the `buf`.

File: pkg/vm/opcode/opcode.go [hunk 3]
```diff
@@ -8,105 +8,41 @@ type Opcode byte
...
+	PUSH9  Opcode = 0x19
+	PUSH10 Opcode = 0x1A
+	PUSH11 Opcode = 0x1B
+	PUSH12 Opcode = 0x1C
+	PUSH13 Opcode = 0x1D
+	PUSH14 Opcode = 0x1E
+	PUSH15 Opcode = 0x1F
+	PUSH16 Opcode = 0x20
+
+	OLDPUSH1 Opcode = 0x51
```
Comment (roman-khimov) [inline pkg/vm/opcode/opcode.go:45]: Looks weird, but let it stay for a while.

File: pkg/core/helper_test.go [hunk 4]
```diff
@@ -550,3 +551,44 @@ func signTx(bc *Blockchain, txs ...*transaction.Transaction) error {
 	}
 	return nil
 }
+
+func getValidatorPass(i int) string {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/helper_test.go:555]: Fixed.

File: pkg/consensus/consensus_test.go [hunk 5]
```diff
@@ -301,7 +301,7 @@ func signTx(t *testing.T, txs ...*transaction.Transaction) {
 		var invoc []byte
 		for _, key := range privNetKeys {
 			signature := key.Sign(data)
-			invoc = append(invoc, append([]byte{byte(opcode.PUSHBYTES64)}, signature...)...)
+			invoc = append(invoc, append([]byte{byte(opcode.PUSHDATA1), 64}, signature...)...)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/consensus/consensus_test.go:304]: Fixed.

File: pkg/consensus/payload.go [hunk 6]
```diff
@@ -192,7 +192,7 @@ func (p *Payload) Sign(key *privateKey) error {
 		return err
 	}
 
-	p.Witness.InvocationScript = append([]byte{byte(opcode.PUSHBYTES64)}, sig...)
+	p.Witness.InvocationScript = append([]byte{byte(opcode.PUSHDATA1), 64}, sig...)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/consensus/payload.go:195]: Fixed.

File: pkg/core/helper_test.go [hunk 7]
```diff
@@ -105,7 +105,7 @@ func newBlock(cfg config.ProtocolConfiguration, index uint32, prev util.Uint256,
 		if len(sig) != 64 {
 			panic("wrong signature length")
 		}
-		invScript = append(invScript, byte(opcode.PUSHBYTES64))
+		invScript = append(invScript, byte(opcode.PUSHDATA1), 64)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/helper_test.go:108]: Fixed.

File: pkg/core/helper_test.go [hunk 8]
```diff
@@ -185,7 +185,7 @@ func newDumbBlock() *block.Block {
 
 func getInvocationScript(data []byte, priv *keys.PrivateKey) []byte {
 	signature := priv.Sign(data)
-	return append([]byte{byte(opcode.PUSHBYTES64)}, signature...)
+	return append([]byte{byte(opcode.PUSHDATA1), 64}, signature...)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/helper_test.go:188]: Fixed.

File: pkg/vm/emit/emit.go (hunk 1 above)
Comment (fyrchik) [reply pkg/vm/emit/emit.go:56]: Fixed.

File: pkg/vm/emit/emit.go (hunk 2 above)
Comment (fyrchik) [reply pkg/vm/emit/emit.go:64]: Fixed.

File: pkg/core/helper_test.go [hunk 9]
```diff
@@ -550,3 +551,44 @@ func signTx(bc *Blockchain, txs ...*transaction.Transaction) error {
...
+	case 2:
+		return "four"
+	case 3:
+		return "three"
+	default:
+		panic("unknown index")
+	}
+}
+
+func TestPrivnetAddresses(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/helper_test.go:556]: Yeah, regenerating will be fun. I will remove this for now.

File: pkg/vm/emit/emit.go [hunk 10]
```diff
@@ -32,20 +33,27 @@ func Bool(w *io.BinWriter, ok bool) {
...
 		val := opcode.Opcode(int(opcode.PUSH1) - 1 + int(i))
 		Opcode(w, val)
 	default:
-		bInt := big.NewInt(i)
-		val := IntToBytes(bInt)
-		Bytes(w, val)
+		buf := intToBytes(big.NewInt(i), make([]byte, 0, 32))
+		// l != 0 becase of switch
+		padSize := byte(8 - bits.LeadingZeros8(byte(len(buf)-1)))
+		Opcode(w, opcode.Opcode(padSize))
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:55]: Why not `w.WriteB()` here? `padSize` is not an opcode anyway.

File: pkg/vm/emit/emit.go [hunk 11]
```diff
@@ -32,20 +33,27 @@ func Bool(w *io.BinWriter, ok bool) {
 	Opcode(w, opcode.PUSHF)
 }
 
+func padRight(s int, buf []byte) []byte {
+	if len(buf) == s {
```
Comment (roman-khimov) [inline pkg/vm/emit/emit.go:37]: `if len(buf) == s` then `buf[:s]` == `buf` and this condition doesn't make much sense. I think we can actually easily get away without `padRight`.

File: pkg/vm/emit/emit.go (hunk 10 above)
Comment (fyrchik) [reply pkg/vm/emit/emit.go:55]: To not violate incapsulation. I know that `8, 16, 32, etc.` are sucessive opcodes starting with `0`, and that's all. May be `opcode.PUSHINT8 + padSize` will be more aproppriate.

Comment (roman-khimov) [reply pkg/vm/emit/emit.go:55]: Yep, it would look a little better.

File: pkg/vm/emit/emit.go (hunk 11 above)
Comment (fyrchik) [reply pkg/vm/emit/emit.go]: It is needed for negative numbers. I have implemented this and added the test.

File: pkg/vm/emit/emit_test.go [hunk 12]
```diff
@@ -35,20 +35,49 @@ func TestEmitInt(t *testing.T) {
...
 		buf := io.NewBufBinWriter()
-		Int(buf.BinWriter, 1000)
+		Int(buf.BinWriter, 1<<28)
+		result := buf.Bytes()
+		assert.Equal(t, 5, len(result))
+		assert.EqualValues(t, opcode.PUSHINT32, result[0])
+		assert.EqualValues(t, 1<<28, BytesToInt(result[1:]).Int64())
+	})
+
+	t.Run("negative 2-byte int with padding", func(t *testing.T) {
```
Comment (roman-khimov) [inline pkg/vm/emit/emit_test.go:73]: 2-byte?

Comment (fyrchik) [reply pkg/vm/emit/emit_test.go]: Fixed, was thinking about normal encoding.

**Pull request merged**